components/CodeWindow.js

```diff
@@ -0,0 +1,280 @@
+import React, { PureComponent } from 'react'
+import * as hljs from 'highlight.js'
+import Spinner from 'react-spinner'
+import ResizeObserver from 'resize-observer-polyfill'
+import debounce from 'lodash.debounce'
+import ms from 'ms'
+
+import WindowControls from '../components/WindowControls'
+import Watermark from '../components/svg/Watermark'
+import CodeMirror from '../lib/react-codemirror'
+import { COLORS, LANGUAGE_MODE_HASH, LANGUAGE_NAME_HASH, DEFAULT_SETTINGS } from '../lib/constants'
+
+class CodeWindow extends PureComponent {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      loading: true,
+      language: props.config.language
+    }
+
+    this.handleLanguageChange = this.handleLanguageChange.bind(this)
+    this.handleTitleBarChange = this.handleTitleBarChange.bind(this)
+    this.codeUpdated = this.codeUpdated.bind(this)
+  }
+
+  componentDidMount() {
+    this.setState({
+      loading: false
+    })
+
+    this.handleLanguageChange(this.props.children)
+
+    const ro = new ResizeObserver(entries => {
+      const cr = entries[0].contentRect
+      this.props.onAspectRatioChange(cr.width / cr.height)
+    })
+    ro.observe(this.exportContainerNode)
+  }
+
+  componentWillReceiveProps(newProps) {
+    // TODO use getDerivedStateFromProps() on React@16.3
+    this.handleLanguageChange(newProps.children, { customProps: newProps })
+  }
+
+  codeUpdated(newCode) {
+    this.handleLanguageChange(newCode)
+    this.props.updateCode(newCode)
+  }
+
+  handleTitleBarChange(newTitle) {
+    this.props.updateTitleBar(newTitle)
+  }
+
+  handleLanguageChange = debounce(
```

Comment:
Can this be extracted to the state container?

Reply:
Yes. But can we extract it later? there are a bunch of spots we can add new subscriptions and containers, but the goal of this PR is mostly to refactor the top level design and let Carbon be exportable on it's own.